
## 股票数据
你已经了解了时间序列数据，下面学习如何将股价数据放入 Pandas 中。
## 读取数据
数据集可能是各种不同的格式。最常见的格式是 [CSV](https://en.wikipedia.org/wiki/Comma-separated_values)。我们将使用“prices.csv”文件作为示例 csv 文件。

In [2]:
with open('prices.csv', 'r') as file:
    prices = file.read()
    
print(prices)

ABC,2017-09-05,163.09,164.24,160.21,162.63,29417590.0,162.49,29414672.0
ABC,2017-09-06,162.85,162.46,159.99,161.13,21131267.0,162.44,21169319.0
ABC,2017-09-07,162.11,162.7,160.65,161.26,21722502.0,161.46,21719856.0
ABC,2017-09-08,160.41,160.89,159.31,158.05,28311012.0,158.26,28305810.0
ABC,2017-09-11,161.09,162.14,159.54,161.29,31075573.0,160.97,31163734.0
ABC,2017-09-12,162.54,164.61,159.52,161.09,70921229.0,160.62,71097150.0
ABC,2017-09-13,160.01,160.51,158.22,159.29,44580353.0,159.07,44260255.0
EFG,2017-09-05,154.45,154.69,153.17,154.52,1270203.0,153.58,1270679.0
EFG,2017-09-06,155.03,155.14,153.89,154.45,1195987.0,154.06,1196107.0
EFG,2017-09-07,154.73,155.36,153.6,155.68,1420730.0,155.6,1409098.0
EFG,2017-09-08,156.01,155.91,154.17,155.86,1438929.0,156.08,1445338.0
EFG,2017-09-11,157.07,157.71,155.93,157.17,1608840.0,156.7,1610357.0
EFG,2017-09-12,155.98,156.72,154.28,156.71,1692197.0,156.86,1687819.0
EFG,2017-09-13,156.4,157.07,155.68,155.54,1211779.0,155.85,1210716.0
XYZ,2017-09

数据提供器将提供关于 CSV 中每个字段的信息。此 csv 包含以下字段：ticker、date、open、high、low、close、volume、adj_close、adj_volume（按顺序排列）。第一行包含以下数据：

- ticker:ABC
- date:2017-09-05
- open:163.09
- high:164.24
- low:160.21
- close:162.63
- volume:29417590.0
- adj_close:162.49
- adj_volume:29414672.0

我们将此数据放入 DataFrame 中。我们将使用[`pd.read_csv`](https://pandas.pydata.org/pandas-docs/version/0.21.1/generated/pandas.read_csv.html) 函数完成此操作。此函数使我们能够根据 CSV 数据生成一个 DataFrame。

In [3]:
import pandas as pd

price_df = pd.read_csv('prices.csv')

price_df

ABC  2017-09-05  163.09  164.24  160.21  162.63  29417590.0  162.49  \
0   ABC  2017-09-06  162.85  162.46  159.99  161.13  21131267.0  162.44   
1   ABC  2017-09-07  162.11  162.70  160.65  161.26  21722502.0  161.46   
2   ABC  2017-09-08  160.41  160.89  159.31  158.05  28311012.0  158.26   
3   ABC  2017-09-11  161.09  162.14  159.54  161.29  31075573.0  160.97   
4   ABC  2017-09-12  162.54  164.61  159.52  161.09  70921229.0  160.62   
5   ABC  2017-09-13  160.01  160.51  158.22  159.29  44580353.0  159.07   
6   EFG  2017-09-05  154.45  154.69  153.17  154.52   1270203.0  153.58   
7   EFG  2017-09-06  155.03  155.14  153.89  154.45   1195987.0  154.06   
8   EFG  2017-09-07  154.73  155.36  153.60  155.68   1420730.0  155.60   
9   EFG  2017-09-08  156.01  155.91  154.17  155.86   1438929.0  156.08   
10  EFG  2017-09-11  157.07  157.71  155.93  157.17   1608840.0  156.70   
11  EFG  2017-09-12  155.98  156.72  154.28  156.71   1692197.0  156.86   
12  EFG  2017-09-13  156.40  157.07  155.68  155.54   1211779.0  155.85   
13  XYZ  2017-09-05   63.90   64.51   63.13   63.95   1738651.0   63.33   
14  XYZ  2017-09-06   63.85   63.65   61.72   62.23   3730110.0   61.95   
15  XYZ  2017-09-07   61.97   61.93   59.47   60.46   6166046.0   60.64   
16  XYZ  2017-09-08   60.36   60.45   58.51   59.35   5173590.0   59.40   
17  XYZ  2017-09-11   60.04   59.92   57.68   58.24   5003322.0   58.02   
18  XYZ  2017-09-12   58.19   59.29   57.89   58.71   3633446.0   58.96   
19  XYZ  2017-09-13   59.01   60.66   58.80   60.33   3571591.0   60.46   

    29414672.0  
0   21169319.0  
1   21719856.0  
2   28305810.0  
3   31163734.0  
4   71097150.0  
5   44260255.0  
6    1270679.0  
7    1196107.0  
8    1409098.0  
9    1445338.0  
10   1610357.0  
11   1687819.0  
12   1210716.0  
13   1733249.0  
14   3725435.0  
15   6191712.0  
16   5174940.0  
17   5001118.0  
18   3635132.0  
19   3583560.0

它会使用 CSV 生成一个 DataFrame，但是认为第一行包含字段名称。我们需要向该函数的参数 `names` 提供一个字段名称列表。

In [4]:
price_df = pd.read_csv('prices.csv', names=['ticker', 'date', 'open', 'high', 'low',
                                             'close', 'volume', 'adj_close', 'adj_volume'])

price_df

ticker        date    open    high     low   close      volume  adj_close  \
0     ABC  2017-09-05  163.09  164.24  160.21  162.63  29417590.0     162.49   
1     ABC  2017-09-06  162.85  162.46  159.99  161.13  21131267.0     162.44   
2     ABC  2017-09-07  162.11  162.70  160.65  161.26  21722502.0     161.46   
3     ABC  2017-09-08  160.41  160.89  159.31  158.05  28311012.0     158.26   
4     ABC  2017-09-11  161.09  162.14  159.54  161.29  31075573.0     160.97   
5     ABC  2017-09-12  162.54  164.61  159.52  161.09  70921229.0     160.62   
6     ABC  2017-09-13  160.01  160.51  158.22  159.29  44580353.0     159.07   
7     EFG  2017-09-05  154.45  154.69  153.17  154.52   1270203.0     153.58   
8     EFG  2017-09-06  155.03  155.14  153.89  154.45   1195987.0     154.06   
9     EFG  2017-09-07  154.73  155.36  153.60  155.68   1420730.0     155.60   
10    EFG  2017-09-08  156.01  155.91  154.17  155.86   1438929.0     156.08   
11    EFG  2017-09-11  157.07  157.71  155.93  157.17   1608840.0     156.70   
12    EFG  2017-09-12  155.98  156.72  154.28  156.71   1692197.0     156.86   
13    EFG  2017-09-13  156.40  157.07  155.68  155.54   1211779.0     155.85   
14    XYZ  2017-09-05   63.90   64.51   63.13   63.95   1738651.0      63.33   
15    XYZ  2017-09-06   63.85   63.65   61.72   62.23   3730110.0      61.95   
16    XYZ  2017-09-07   61.97   61.93   59.47   60.46   6166046.0      60.64   
17    XYZ  2017-09-08   60.36   60.45   58.51   59.35   5173590.0      59.40   
18    XYZ  2017-09-11   60.04   59.92   57.68   58.24   5003322.0      58.02   
19    XYZ  2017-09-12   58.19   59.29   57.89   58.71   3633446.0      58.96   
20    XYZ  2017-09-13   59.01   60.66   58.80   60.33   3571591.0      60.46   

    adj_volume  
0   29414672.0  
1   21169319.0  
2   21719856.0  
3   28305810.0  
4   31163734.0  
5   71097150.0  
6   44260255.0  
7    1270679.0  
8    1196107.0  
9    1409098.0  
10   1445338.0  
11   1610357.0  
12   1687819.0  
13   1210716.0  
14   1733249.0  
15   3725435.0  
16   6191712.0  
17   5174940.0  
18   5001118.0  
19   3635132.0  
20   3583560.0

## DataFrame 计算
将数据放入 DataFrame 后，我们可以对其进行运算。我们使用 [`DataFrame.median`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.median.html) 函数算出每支股票的中位数。

In [5]:
price_df.median()

open              155.98
high              155.91
low               154.17
close             155.68
volume        3730110.00
adj_close         155.85
adj_volume    3725435.00
dtype: float64

不正确。这些值是整个股票备选投资组合的中位数。我们将使用 [`DataFrame.groupby`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.groupby.html) 函数获取每支股票的均值。

In [6]:
price_df.groupby('ticker').median()

open    high     low   close      volume  adj_close  adj_volume
ticker                                                                   
ABC     162.11  162.46  159.54  161.13  29417590.0     160.97  29414672.0
EFG     155.98  155.91  154.17  155.68   1420730.0     155.85   1409098.0
XYZ      60.36   60.66   58.80   60.33   3730110.0      60.46   3725435.0

这就是我们要计算的值。但是，我们不希望每次进行运算都运行 `groupby` 函数。我们可通过 `price_df_ticker_groups = price_df.groupby('ticker')` 保存该 GroupBy 对象。这限制我们只能对 GroupBy 对象进行运算。还可以使用 [`GroupBy.apply`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.core.groupby.GroupBy.apply.html)，但是会降低性能。真正的问题是数据的表示形式。

In [8]:
price_df_ticker_groups = price_df.groupby('ticker')
price_df_ticker_groups

In [7]:
price_df.iloc[:16]

ticker        date    open    high     low   close      volume  adj_close  \
0     ABC  2017-09-05  163.09  164.24  160.21  162.63  29417590.0     162.49   
1     ABC  2017-09-06  162.85  162.46  159.99  161.13  21131267.0     162.44   
2     ABC  2017-09-07  162.11  162.70  160.65  161.26  21722502.0     161.46   
3     ABC  2017-09-08  160.41  160.89  159.31  158.05  28311012.0     158.26   
4     ABC  2017-09-11  161.09  162.14  159.54  161.29  31075573.0     160.97   
5     ABC  2017-09-12  162.54  164.61  159.52  161.09  70921229.0     160.62   
6     ABC  2017-09-13  160.01  160.51  158.22  159.29  44580353.0     159.07   
7     EFG  2017-09-05  154.45  154.69  153.17  154.52   1270203.0     153.58   
8     EFG  2017-09-06  155.03  155.14  153.89  154.45   1195987.0     154.06   
9     EFG  2017-09-07  154.73  155.36  153.60  155.68   1420730.0     155.60   
10    EFG  2017-09-08  156.01  155.91  154.17  155.86   1438929.0     156.08   
11    EFG  2017-09-11  157.07  157.71  155.93  157.17   1608840.0     156.70   
12    EFG  2017-09-12  155.98  156.72  154.28  156.71   1692197.0     156.86   
13    EFG  2017-09-13  156.40  157.07  155.68  155.54   1211779.0     155.85   
14    XYZ  2017-09-05   63.90   64.51   63.13   63.95   1738651.0      63.33   
15    XYZ  2017-09-06   63.85   63.65   61.72   62.23   3730110.0      61.95   

    adj_volume  
0   29414672.0  
1   21169319.0  
2   21719856.0  
3   28305810.0  
4   31163734.0  
5   71097150.0  
6   44260255.0  
7    1270679.0  
8    1196107.0  
9    1409098.0  
10   1445338.0  
11   1610357.0  
12   1687819.0  
13   1210716.0  
14   1733249.0  
15   3725435.0

你能发现问题吗？花点时间找找问题。

问题出在行 [6, 7] 和 [13, 14] 之间

In [9]:
price_df.iloc[[6, 7, 13, 14]]

ticker        date    open    high     low   close      volume  adj_close  \
6     ABC  2017-09-13  160.01  160.51  158.22  159.29  44580353.0     159.07   
7     EFG  2017-09-05  154.45  154.69  153.17  154.52   1270203.0     153.58   
13    EFG  2017-09-13  156.40  157.07  155.68  155.54   1211779.0     155.85   
14    XYZ  2017-09-05   63.90   64.51   63.13   63.95   1738651.0      63.33   

    adj_volume  
6   44260255.0  
7    1270679.0  
13   1210716.0  
14   1733249.0

所有股票的数据堆叠到一起了。我们用二维空间表示三维数据。我们使用 Panda 的 Panel 解决了此问题，Panel [已被弃用](https://pandas.pydata.org/pandas-docs/version/0.21.0/dsintro.html#deprecate-panel)。Pandas 文档建议我们使用 [MultiIndex](https://pandas.pydata.org/pandas-docs/version/0.21/advanced.html) 或[xarray](http://xarray.pydata.org/en/stable/)。[MultiIndex](https://pandas.pydata.org/pandas-docs/version/0.21/advanced.html) 依然不能解决我们的问题，因为数据依然用二维空间表示。[xarray](http://xarray.pydata.org/en/stable/) 能够存储三维数据，但是 Finance 使用的是 Pandas，所以我们将继续使用此库。学完此课程后，建议看看 [xarray](http://xarray.pydata.org/en/stable/)。

如何在 Pandas 中使用三维数据？我们可以将每个第三维拆分成它自己的二维 DataFrame。以下面的数组为例：
```
[
    [
        [ 0,  1],
        [ 2,  3],
        [ 4,  5]
    ],[
        [ 6,  7],
        [ 8,  9],
        [10, 11]
    ],[
        [12, 13],
        [14, 15],
        [16, 17]
    ],[
        [18, 19],
        [20, 21],
        [22, 23]
    ]
]    
```

我们需要将它拆分成以下两个二维数组：
```
[
    [0, 2, 4],
    [6, 8, 10],
    [12, 14, 16],
    [18, 20, 22]
] 
[
    [1, 3, 5],
    [7, 8, 11],
    [13, 15, 17],
    [19, 21, 23]
] 
```

对于此示例来说，第三维是“open”、“high”、“low”、“close”、“volume”、“adj_close”和“adj_volume”。我们将使用 [`DataFrame.pivot`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.pivot.html) 函数生成这些 DataFrame。

In [10]:
open_prices = price_df.pivot(index='date', columns='ticker', values='open')
high_prices = price_df.pivot(index='date', columns='ticker', values='high')
low_prices = price_df.pivot(index='date', columns='ticker', values='low')
close_prices = price_df.pivot(index='date', columns='ticker', values='close')
volume = price_df.pivot(index='date', columns='ticker', values='volume')
adj_close_prices = price_df.pivot(index='date', columns='ticker', values='adj_close')
adj_volume = price_df.pivot(index='date', columns='ticker', values='adj_volume')

open_prices

ticker         ABC     EFG    XYZ
date                             
2017-09-05  163.09  154.45  63.90
2017-09-06  162.85  155.03  63.85
2017-09-07  162.11  154.73  61.97
2017-09-08  160.41  156.01  60.36
2017-09-11  161.09  157.07  60.04
2017-09-12  162.54  155.98  58.19
2017-09-13  160.01  156.40  59.01

这样就可以获得所有开盘价、最高价、最低价等的 DataFrame。我们最终要计算的是每支股票的均值。

In [11]:
open_prices.mean()

ticker
ABC    161.728571
EFG    155.667143
XYZ     61.045714
dtype: float64

还可以通过转置获得每个日期的均值。

In [19]:
import numpy as np
a = np.arange(120).reshape(2,3,4,5)
b= a.T

In [20]:
b.shape

(5, 4, 3, 2)

#### 复习：
- 多维数组的transpose,a[i][j][k]的转置是a[k][j][i]

#### 注意：
- 使用pivot方法，相当于形成了二维数组，也就是‘index’和‘column’指定的最高维，和次高维，而数值作为最低维。
- 以下采用的transpose方法，实际上是，数值不变，也就是说最低维不变，只是‘index’和‘column’两个维度进行转置。

In [22]:
open_prices.T

date    2017-09-05  2017-09-06  2017-09-07  2017-09-08  2017-09-11  \
ticker                                                               
ABC         163.09      162.85      162.11      160.41      161.09   
EFG         154.45      155.03      154.73      156.01      157.07   
XYZ          63.90       63.85       61.97       60.36       60.04   

date    2017-09-12  2017-09-13  
ticker                          
ABC         162.54      160.01  
EFG         155.98      156.40  
XYZ          58.19       59.01

In [12]:
open_prices.T.mean()

date
2017-09-05    127.146667
2017-09-06    127.243333
2017-09-07    126.270000
2017-09-08    125.593333
2017-09-11    126.066667
2017-09-12    125.570000
2017-09-13    125.140000
dtype: float64

日期在索引中、股票代号在列中或反过来都没关系。始终相互转置。由于我们将对日期进行很多运算，所以在这门课程中，我们将把日期放入索引中，并把股票代号放入列中。
## 小测验
下面看看你能否运用所学的知识。请实现 `csv_to_close` 函数，它将接受文件路径和 `csv_filename`，并输出收盘价二维数组。你可以认为 `csv_to_close` 使用的 CSV 文件具有和“prices.csv”相同的名称并且字段顺序一样。

为了帮助你完成小测验，我们提供了单元测试供你测试函数实现情况。对于这道练习，我们将使用 `quiz_tests` 模块中的函数 `test_csv_to_close` 测试 `csv_to_close`。

In [25]:
import quiz_tests


def csv_to_close(csv_filepath, field_names):
    """Reads in data from a csv file and produces a DataFrame with close data.
    
    Parameters
    ----------
    csv_filepath : str
        The name of the csv file to read
    field_names : list of str
        The field names of the field in the csv file

    Returns
    -------
    close : DataFrame
        Close prices for each ticker and date
    """
    
    import pandas as pd
    prices_df = pd.read_csv(csv_filepath,names=['ticker', 'date', 'open', 'high', 'low',
                                    'close', 'volume', 'adj_close', 'adj_volume'])
    close_prices = prices_df.pivot(index='date',columns='ticker',values='close')
    
    return close_prices


quiz_tests.test_csv_to_close(csv_to_close)

Tests Passed


## 小测验解答
如果你遇到问题，请在[此处](stock_data_solution.ipynb)查看解答。